node/silkworm/downloader/internals/header_only_state.hpp

```diff
@@ -0,0 +1,102 @@
+/*
+Copyright 2020-2021 The Silkworm Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+#ifndef SILKWORM_HEADER_ONLY_STATE_HPP
+#define SILKWORM_HEADER_ONLY_STATE_HPP
+
+#include <silkworm/state/state.hpp>
+#include "chain_elements.hpp"
+#include "types.hpp"
+
+namespace silkworm {
+
+class HeaderOnlyChainStateBase : public State {
```

Comment:
Can we do it the other way round? I mean have something like `HeaderOnlyChainState` interface with just `read_header` & `insert_header`, and inherit `State` from `HeaderOnlyChainState`.

Reply:
It would be better! Can i change State class?

Reply:
> It would be better! Can I change State class?

Sure, you can change anything in our code base :)

Reply:
Thanks! But there is another issue: IConsensusEngine has this method:
```
ValidationResult ConsensusEngineBase::validate_block_header(const BlockHeader& header, State& state,
                                                            bool with_future_timestamp_check)
```
that we can change using HeaderOnlyChainState in place of State; but it has also, for example, this method:

`virtual ValidationResult pre_validate_block(const Block& block, State& state) = 0;`

where we cannot use HeaderOnlyChainState in place of State, beacuse it require a full State for block validation. So we need also to introduce some kind of HeaderOnlyConsensusEngine or HeaderBasedConsensusEngine as a base class for IConsensusEngine.

Is it ok?

Reply:
`ConsensusEngineBase::pre_validate_block` only reads headers and bodies from the state. Perhaps we can have `BlockState` interface (instead of `HeaderOnlyChainState`) with `read_header`,  `insert_header`, and `read_body` and change `IConsensusEngine::pre_validate_block` to take `BlockState` instead of `State`.